gr-blocks/lib/message_debug_impl.cc

```diff
@@ -20,16 +20,36 @@
 namespace gr {
 namespace blocks {
 
-message_debug::sptr message_debug::make()
+message_debug::sptr message_debug::make(bool en_uvec)
 {
-    return gnuradio::get_initial_sptr(new message_debug_impl());
+    return gnuradio::get_initial_sptr(new message_debug_impl(en_uvec));
 }
 
+message_debug_impl::message_debug_impl(bool en_uvec)
+    : block("message_debug", io_signature::make(0, 0, 0), io_signature::make(0, 0, 0)),
+      d_en_uvec(en_uvec)
+{
+    message_port_register_in(pmt::mp("print"));
+    set_msg_handler(pmt::mp("print"), boost::bind(&message_debug_impl::print, this, _1));
+
+    message_port_register_in(pmt::mp("store"));
+    set_msg_handler(pmt::mp("store"), boost::bind(&message_debug_impl::store, this, _1));
+
+    message_port_register_in(pmt::mp("print_pdu"));
+    set_msg_handler(pmt::mp("print_pdu"),
+                    boost::bind(&message_debug_impl::print_pdu, this, _1));
+}
+
+message_debug_impl::~message_debug_impl() {}
+
 void message_debug_impl::print(pmt::pmt_t msg)
 {
-    std::cout << "******* MESSAGE DEBUG PRINT ********\n";
-    pmt::print(msg);
-    std::cout << "************************************\n";
+    std::stringstream sout;
+
+    sout << "******* MESSAGE DEBUG PRINT ********" << std::endl;
+    sout << pmt::write_string(msg) << std::endl;
+    sout << "************************************" << std::endl;
+    std::cout << sout.str();
```

Comment:
What's the intermediate `stringstream` supposed to accomplish? Writing to stdout is not atomic either way.

Reply:
It’s not atomic but it’s much much better in terms of getting associated printouts together  without mixing them. Intermediate stringstreams are used in other in tree blocks like the tag debug.

You have lots of good comments, I’ll get to the rest soon thanks!

Reply:
This doesn't prevent clashing output. You're saying though that in practice it improves it?